design-documents/allowed-http-methods-for-actions.md

```diff
@@ -0,0 +1,56 @@
+### Overview
+
+To improve security and logistics we need to allow limiting Actions to processing only requests with certain HTTP methods and add those limitations to as many existing Actions as possible.
+There are many vulnerabilities caused by actions processing both GET and POST requests and thus allowing bypassing security validations like form key validation. Also limiting actions to processing only requests with certain methods would serve as self-documentation for Action classes and improve consistency of server side for client code and functional tests.
+
+### Theory of Operation
+
+Since we use by-convention routing it would be inconsistent to defined actions’ allowed HTTP methods in a config file, so it’s better to create marker interfaces like these:
+
+>interface HttpGetActionInterface extends ActionInterface
+>
+>{
+>
+>}
+>
+>
+>interface HttpPostActionInterface extends ActionInterface
+>
+>{
+>
+>}
+>
+>
+>interface HttpPutActionInterface extends ActionInterface
+>
+>{
+>
+>}
+
+etc.
+
+One interface per each HTTP method (using list of HTTP methods from \Zend\Http\Request::METHOD_* constants). We will have an exentable map *interface* => *HTTP method* passed to the validator to allow custom HTTP methods to be processed.
+
+Request method will be validated in the FrontController before executing a matched action, somewhat like this:
+
+>foreach ($interfaceMap as $interface => $httpMethod) {
+>
+>&nbsp;&nbsp;&nbsp;&nbsp;if ($actionInstance instanceof $interface && $request->getMethod() !== $httpMethod) {
+>
+>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;throw new NotFoundException(__(‘Page not found.’));
+>
+>&nbsp;&nbsp;&nbsp;&nbsp;}
+>
+>}
+
+ 
+To limit already existing Actions and get developers to know these new interfaces a temporary event listener will be added for controller_action_postdispatch event to log HTTP methods used with actions and then the full suite of functional tests (MTF and MFTF) would be ran to get information on as much actions as possible.
+
+After corresponding Http*Method*ActionInterface will be added to every Action class we have information on. If an action is logged to be accessed only by GET methods it will allow only GET, if only by POST – allowing only POST etc. If by more than one methods – the Action class will not be updated.
```

Comment:
> If by more than one methods – the Action class will not be updated

Why not make it `implement` multiple interfaces? Or if it's not clear why it really supports multiple methods, create a task to review it (as it can be a security issue).

Reply:
because we cannot be sure that an action actually expects multiple request methods or a test is invalid without investigation and if too many cases like that would be found it can take too match time to investigate them. I think I can write something like this: "If by more than one methods – the Action class will be updated to implement multiple interfaces or ignored." Would that be ok?

Reply:
I agree with Olga and would propose for now agree that we will need to review each case and discuss one more time if turns out that there are too many cases that need to be reviewed. Or you already know?